include/boost/numeric/ublas/tensor/tensor.hpp

```diff
@@ -7,615 +7,193 @@
-        public detail::tensor_expression< basic_tensor<T>,basic_tensor<T> >
-{
-
-    using self_type                 = basic_tensor<T>;
-
-public:
-
-    template<class derived_type>
-    using tensor_expression_type    = detail::tensor_expression<self_type,derived_type>;
-
-    template<class derived_type>
-    using matrix_expression_type    = matrix_expression<derived_type>;
-
-    template<class derived_type>
-    using vector_expression_type    = vector_expression<derived_type>;
-
-    using super_type                = tensor_expression_type<self_type>;
-
-    using array_type                = typename tensor_traits<T>::container_type;
-    using layout_type               = typename tensor_traits<T>::layout_type;
-
-
-    using size_type                 = typename array_type::size_type;
-    using difference_type           = typename array_type::difference_type;
-    using value_type                = typename array_type::value_type;
-
-    using reference                 = typename array_type::reference;
-    using const_reference           = typename array_type::const_reference;
-
-    using pointer                   = typename array_type::pointer;
-    using const_pointer             = typename array_type::const_pointer;
-
-    using iterator                  = typename array_type::iterator;
-    using const_iterator            = typename array_type::const_iterator;
-
-    using reverse_iterator          = typename array_type::reverse_iterator;
-    using const_reverse_iterator    = typename array_type::const_reverse_iterator;
-
-    using tensor_temporary_type     = self_type;
-    using storage_category          = dense_tag;
-    using container_tag             = typename tensor_traits<T>::container_tag;
+namespace boost::numeric::ublas{
 
-    using extents_type              = typename tensor_traits<T>::extents_type;
-    using strides_type              = strides_t<extents_type,layout_type>;
+    namespace layout{
 
-    using matrix_type               = matrix<value_type,layout_type, std::vector<value_type> >;
-    using vector_type               = vector<value_type, std::vector<value_type> >;
+        template<typename...>
+        struct first_order;
+        
+        template<>
+        struct first_order<>{
+            
+            template<typename ExtentsType>
+            struct strides{
+                using extents_type = ExtentsType;
+                using layout_type = ::boost::numeric::ublas::first_order;
+                using strides_type = strides_t<ExtentsType,layout_type>;
+            };
+            
+        };
+        
+        template<typename ExtentsType>
+        struct first_order<ExtentsType>{
+            
+            template<typename>
+            struct strides{
+                using extents_type = ExtentsType;
+                using layout_type = ::boost::numeric::ublas::first_order;
+                using strides_type = strides_t<ExtentsType,layout_type>;
+            };
+
+        };
```

Comment:
`layout` and `extents` are orthogonal concepts where the definition of layout should not depend on the type of extents. 
the implementation does not reflect that. Also, I recommend not to have `strides` struct inside a `layout`.

Reply:
If the decouple them then I have to add a new template parameter and construction of strides depends on the layout so they are interdependent on each other. I can separate them if you want.

Reply:
yes - please. but why was this necessary at all?

Reply:
to provide the construction of tensor with different strides and remove the redundancy to provide the layout once and again when providing the strides because of the interface we discussed.

Reply:
removed it and added a new header file 'layout.hpp' and template argument. Made a few changes to the `type_traits_strides` to adjust with the template argument `StrideType`.